Salesforce usability improvements

GTFalcao opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Enhanced management of Salesforce records with structured property management and validation.
	- Introduction of improved property management for Salesforce opportunities, accounts, cases, comments, tasks, and users.
	- New user properties module for better management of user data.

- **Bug Fixes**
	- Updated property names for clarity and consistency in actions.
	- Improved error handling and messaging for user feedback across various actions.

- **Documentation**
	- Updated description fields to provide clearer explanations and direct links to relevant documentation.
	- Streamlined documentation for new features and changes to ensure users are informed of updates.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

GTFalcao commented on components/salesforce_rest_api/sources/common.mjs:
```diff
@@ -55,8 +94,8 @@ export default {
     },
-    processEvent() {
-      throw new Error("processEvent is not implemented");
+    processTimerEvent() {
+      throw new Error("processTimerEvent is not implemented");
```
This object is meant to be extended, akin to an abstract class. Hence, it intentionally does not implement the `processTimerEvent` method, so that objects that extend from it have to implement the method (or throw an error if they do not).

GTFalcao commented on components/salesforce_rest_api/sources/common-webhook-methods.mjs:
```diff
@@ -0,0 +1,71 @@
+    this.db.set("webhookData", webhookData);
+  },
+  _isValidSource(event) {
+    const webhookToken = event.headers["x-webhook-token"];
+    const secretToken = this._getSecretToken();
+    return webhookToken === secretToken;
+  },
+  processWebhookEvent(event) {
+    const { body } = event;
+    const meta = this.generateWebhookMeta(event);
+    this.$emit(body, meta);
+  },
+  generateWebhookMeta() {
+    throw new Error("generateWebhookMeta is not implemented");
+  },
+  getEventType() {
+    throw new Error("getEventType is not implemented");
+  },
+  /**
+   * This method returns the fields in the SObject type (e.g. Account, Lead, etc.) that the event
+   * source should listen for updates to. This base implementation returns `undefined`, to not
+   * necessitate any specific fields to be updated.
+   *
+   * @returns the fields in the SObject type for which to receive updates
+   */
+  getFieldsToCheck() {
+    return undefined;
+  },
+  /**
+   * This method returns whether the event source should listen for updates where `all` the fields
+   * in the SObject are updated, or when `any` of them are. This base implementation returns
+   * `undefined` to use to client's default `fieldToCheckMode` (`any`).
+   *
+   * @returns whether the webhook should receive events when `all` the fields to check are
+   * updated, or when `any` of them are
+   */
+  getFieldsToCheckMode() {
+    return undefined;
+  },
+};
```
This object is meant to be extended, akin to an abstract class. Hence, it intentionally does not implement these methods, so that objects that extend from it have to implement the methods (or throw an error if they do not).

GTFalcao commented on components/salesforce_rest_api/actions/common/base-create-update.mjs:
```diff
@@ -0,0 +1,96 @@
+      return {};
+    },
+    getAdditionalFields,
+    formatDateTimeProps(props = {}) {
+      // https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/intro_valid_date_formats.htm
+      return Object.fromEntries(Object.entries(props).map(([
+        key,
+        value,
+      ]) => {
+        const numValue = Number(value);
+        const date = new Date(isNaN(numValue)
+          ? value
+          : numValue);
+        if (isNaN(date.valueOf())) {
+          throw new ConfigurationError(`Invalid date format for prop \`${key}\`. Please provide a valid date format.`);
+        }
+        return [
+          key,
+          date.toISOString(),
+        ];
+      }));
```
Suggestion implemented, thanks.

luancazarine reviewed (dismissed):
Hi @GTFalcao, LGTM! Ready for QA!

GTFalcao commented:
- Removed `DoNotCall` from Contact, `HasOptedOutOfEmail` from Lead, `ContractId` from Opportunity, 'IsCustomerPortal' from Account;
    - Even though these fields are listed in the API docs, they are not accepted and also not present in the describe call;
- Removed `CurrentStatus` from User - it was marked as deprecated in the docs;
- Fixed date/time prop parsing to properly ignore props that had no value set;
- Fixed 'Create Event' and 'Delete Opportunity' errors;
- Fixed 'SOSL Search' summary;

I could not reproduce the 'Delete Record' issue.

I did not modify 'Insert Blob Data' in this update - I will look into it with Michael as this was potentially going to be deprecated in favor of another type of file handling.

GTFalcao commented:
- Fixed error in base64 parsing of attachment files;
- Changed Case OwnerId from Contact to User (the documentation explicitly says Contact, but apparently the API expects a User);
- Fixed "User Role ID" selecting wrong records;
- Removed "IsPublished" from CaseComment's props;
- Removed "IsPartner" from Account's props;
- Fixed "Update Opportunity" summary;
- The error in "Create Task" is a business rule error (of props incompatible with each other). The "Recurring" prop probably needs to be true for "Recurrence Time Zone" to be used, among others. I had other business rule errors too with other props, I believe this is something users will not face as much as these rules probably make more sense in real world scenarios than in testing. This was my prop setup for a successful run, tweaking props from your inputs:
```
IsRecurrence: true
TaskSubtype: Email
Description: Task from Pipedream 1432
Priority: Normal
CallDisposition: we'll call back
CallDurationInSeconds: 30
CallObject: Test caller
CallType: Inbound
IsReminderSet: true
RecurrenceDayOfWeekMask: [ 2, 4, 16, 8 ]
RecurrenceEndDateOnly: 2024-08-09
RecurrenceStartDateOnly: 2024-08-05
RecurrenceTimeZoneSidKey: Asia/Bangkok
RecurrenceType: RecursEveryWeekday
ReminderDateTime: 2024-08-10
Status: Completed
Subject: Send Letter
```

GTFalcao commented:
/approve
